File: jafu/src/main/java/org/springframework/fu/jafu/webflux/WebFluxServerDsl.java
package org.springframework.fu.jafu.webflux;

import static org.springframework.beans.factory.support.BeanDefinitionReaderUtils.uniqueBeanName;

import java.util.function.Consumer;
import java.util.function.Supplier;

import org.springframework.beans.factory.config.BeanDefinitionCustomizer;
import org.springframework.beans.factory.support.BeanDefinitionReaderUtils;
import org.springframework.boot.WebApplicationType;
import org.springframework.boot.autoconfigure.web.ResourceProperties;
import org.springframework.boot.autoconfigure.web.ServerProperties;
import org.springframework.boot.autoconfigure.web.reactive.FormCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.JacksonJsonCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.MultipartCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.ProtobufCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.ReactiveWebServerInitializer;
import org.springframework.boot.autoconfigure.web.reactive.ResourceCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.StringCodecInitializer;
import org.springframework.boot.autoconfigure.web.reactive.WebFluxProperties;
import org.springframework.boot.web.embedded.netty.NettyReactiveWebServerFactory;
import org.springframework.boot.web.reactive.server.ConfigurableReactiveWebServerFactory;
import org.springframework.context.ApplicationContextInitializer;
import org.springframework.context.support.GenericApplicationContext;
import org.springframework.fu.jafu.AbstractDsl;
import org.springframework.fu.jafu.web.JacksonDsl;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.RouterFunctions;
import org.springframework.web.server.WebFilter;

/**
 * Jafu DSL for WebFlux server.
 *
 * This DSL to be used with {@link org.springframework.fu.jafu.Jafu#application(WebApplicationType, java.util.function.Consumer)}
 * using a {@link WebApplicationType#REACTIVE} parameter configures a
 * <a href="https://docs.spring.io/spring/docs/current/spring-framework-reference/web-reactive.html#spring-webflux"></a>WebFlux server</a>.
 *
 * When no codec is configured, {@code String} and {@code Resource} ones are configured by default.
 * When a {@code codecs} block is declared, the one specified are configured by default.
 *
 * You can chose the underlying engine via the {@link WebFluxServerDsl#engine(ConfigurableReactiveWebServerFactory)} parameter.
 *
 * Required dependencies can be retrieve using {@code org.springframework.boot:spring-boot-starter-webflux}.
 *
 * @see org.springframework.fu.jafu.Jafu#application(WebApplicationType, java.util.function.Consumer)
 * @see WebFluxServerDsl#codecs(Consumer)
 * @see WebFluxServerDsl#mustache()
 * @author Sebastien Deleuze
 */
public class WebFluxServerDsl extends AbstractDsl {

	private final Consumer<WebFluxServerDsl> dsl;

	private ServerProperties serverProperties = new ServerProperties();

	private ResourceProperties resourceProperties = new ResourceProperties();

	private WebFluxProperties webFluxProperties = new WebFluxProperties();

	private boolean codecsConfigured = false;

	private int port = 8080;

	private ConfigurableReactiveWebServerFactory engine = null;

	WebFluxServerDsl(Consumer<WebFluxServerDsl> dsl) {
		super();
		this.dsl = dsl;
	}

	public static ApplicationContextInitializer<GenericApplicationContext> webFlux() {
		return new WebFluxServerDsl(dsl -> {});
	}

	public static ApplicationContextInitializer<GenericApplicationContext> webFlux(Consumer<WebFluxServerDsl> dsl) {
		return new WebFluxServerDsl(dsl);
	}

	/**
	 * Define the listening port of the webFlux.
	 */
	public WebFluxServerDsl port(int port) {
		this.port = port;
		return this;
	}

	/**
	 * Define the underlying engine used.
	 */
	public WebFluxServerDsl engine(ConfigurableReactiveWebServerFactory engine) {
		this.engine = engine;
		return this;
	}

	/**
	 * Configure routes via {@link RouterFunctions.Builder}.
	 * @see org.springframework.fu.jafu.BeanDefinitionDsl#bean(Class, BeanDefinitionCustomizer...)
	 */
	public WebFluxServerDsl router(Consumer<RouterFunctions.Builder> routerDsl) {
		RouterFunctions.Builder builder = RouterFunctions.route();
		context.registerBean(BeanDefinitionReaderUtils.uniqueBeanName(RouterFunction.class.getName(), context), RouterFunction.class, () -> {
			routerDsl.accept(builder);
			return builder.build();
		});
		return this;
	}

	/**
	 * Configure codecs via a [dedicated DSL][WebFluxServerCodecDsl].
	 * @see WebFluxServerCodecDsl#jackson
	 */
	public WebFluxServerDsl codecs(Consumer<WebFluxServerCodecDsl> init) {
		new WebFluxServerCodecDsl(init).initialize(context);
		this.codecsConfigured = true;
		return this;
	}

	/**
	 * Define a request filter for this webFlux
	 */
	public WebFluxServerDsl filter(Class<? extends WebFilter> clazz) {
		context.registerBean(uniqueBeanName(clazz.getName(), context), clazz);
		return this;
	}

	/**
	 * @see #mustache(Consumer)
	 */
	public WebFluxServerDsl mustache() {
		return mustache(dsl -> {});
	}

	/**
	 * Configure Mustache view resolver.
	 *
	 * Require {@code org.springframework.boot:spring-boot-starter-mustache} dependency.
	 */
	public WebFluxServerDsl mustache(Consumer<MustacheDsl> dsl) {
		new MustacheDsl(dsl).initialize(context);
		return this;
	}

	/**
	 * Enable an external codec.
	 */
	@Override
	public WebFluxServerDsl enable(ApplicationContextInitializer<GenericApplicationContext> dsl) {
		return (WebFluxServerDsl) super.enable(dsl);
	}

	@Override
	public void initialize(GenericApplicationContext context) {
		super.initialize(context);
		this.dsl.accept(this);
		if (engine == null) {
			engine = new NettyDelegate().get();
		}
		engine.setPort(port);

		if (!codecsConfigured) {
			new StringCodecInitializer(false, false).initialize(context);
			new ResourceCodecInitializer(false).initialize(context);
		}
		if (context.containsBeanDefinition("webHandler")) {
			throw new IllegalStateException("Only one webFlux per application is supported");
		}
		new ReactiveWebServerInitializer(serverProperties, resourceProperties, webFluxProperties, engine).initialize(context);

	}

	/**
	 * Jafu DSL for WebFlux server codecs.
	 */
	static public class WebFluxServerCodecDsl extends AbstractDsl {

		private final Consumer<WebFluxServerCodecDsl> dsl;

		WebFluxServerCodecDsl(Consumer<WebFluxServerCodecDsl> dsl) {
			this.dsl = dsl;
		}

		@Override
		public WebFluxServerCodecDsl enable(ApplicationContextInitializer<GenericApplicationContext> dsl) {
			return (WebFluxServerCodecDsl) super.enable(dsl);
		}

		@Override
		public void initialize(GenericApplicationContext context) {
			super.initialize(context);
			this.dsl.accept(this);
		}

		/**
		 * Enable {@link org.springframework.core.codec.CharSequenceEncoder} and {@link org.springframework.core.codec.StringDecoder} for all media types
		 */
		public WebFluxServerCodecDsl string() {
			new StringCodecInitializer(false, false).initialize(context);
			return this;
		}

		/**
		 * Enable {@link org.springframework.core.codec.CharSequenceEncoder} and {@link org.springframework.core.codec.StringDecoder}
		 */
		public WebFluxServerCodecDsl string(boolean textPlainOnly) {
			new StringCodecInitializer(false, textPlainOnly).initialize(context);
			return this;
		}

		/**
		 * Enable {@link org.springframework.http.codec.ResourceHttpMessageWriter} and {@link org.springframework.core.codec.ResourceDecoder}
		 */
		public WebFluxServerCodecDsl resource() {
			new ResourceCodecInitializer(false).initialize(context);
			return this;
		}

		/**
		 * Enable {@link org.springframework.http.codec.protobuf.ProtobufEncoder} and {@link org.springframework.http.codec.protobuf.ProtobufDecoder}
		 *
		 * This codec requires Protobuf 3 or higher with the official `com.google.protobuf:protobuf-java` dependency, and
		 * supports `application/x-protobuf` and `application/octet-stream`.
		 */
		public WebFluxServerCodecDsl protobuf() {
			new ProtobufCodecInitializer(false).initialize(context);
			return this;
		}

		/**
		 * Enable {@link org.springframework.http.codec.FormHttpMessageWriter} and {@link org.springframework.http.codec.FormHttpMessageReader}
		 */
		public WebFluxServerCodecDsl form() {
			new FormCodecInitializer(false).initialize(context);
			return this;
		}

		/**
		 * Enable {@link org.springframework.http.codec.multipart.MultipartHttpMessageWriter} and
		 * {@link org.springframework.http.codec.multipart.MultipartHttpMessageReader}
		 *
		 * This codec requires Synchronoss NIO Multipart library via  the {@code org.synchronoss.cloud:nio-multipart-parser} dependency.
		 */
		public WebFluxServerCodecDsl multipart() {
			new MultipartCodecInitializer(false).initialize(context);
			return this;
		}

		/**
		 * @see #jackson(Consumer)
		 */
		public WebFluxServerCodecDsl jackson() {
			return jackson(dsl -> {});
		}

		/**
		 * Register an `ObjectMapper` bean and configure a [Jackson](https://github.com/FasterXML/jackson)
		 * JSON codec on WebFlux server via a [dedicated DSL][JacksonDsl].
		 *
		 * Required dependencies can be retrieve using `org.springframework.boot:spring-boot-starter-json`
		 * (included by default in `spring-boot-starter-webflux`).
		 */
		public WebFluxServerCodecDsl jackson(Consumer<JacksonDsl> dsl) {
			new JacksonDsl(false, dsl).initialize(context);
			new JacksonJsonCodecInitializer(false).initialize(context);
			return this;
		}
	}

	private static class NettyDelegate implements Supplier<ConfigurableReactiveWebServerFactory> {
		@Override
		public ConfigurableReactiveWebServerFactory get() {
			return new NettyReactiveWebServerFactory();
		}
	}

}

File: kofu/src/main/kotlin/org/springframework/fu/kofu/webflux/WebFluxServerDsl.kt
package org.springframework.fu.kofu.webflux

import org.springframework.beans.factory.ObjectProvider
import org.springframework.beans.factory.getBeanProvider
import org.springframework.beans.factory.support.BeanDefinitionReaderUtils.uniqueBeanName
import org.springframework.boot.autoconfigure.web.ResourceProperties
import org.springframework.boot.autoconfigure.web.ServerProperties
import org.springframework.boot.autoconfigure.web.reactive.*
import org.springframework.boot.web.embedded.jetty.JettyReactiveWebServerFactory
import org.springframework.boot.web.embedded.netty.NettyReactiveWebServerFactory
import org.springframework.boot.web.embedded.tomcat.TomcatReactiveWebServerFactory
import org.springframework.boot.web.embedded.undertow.UndertowReactiveWebServerFactory
import org.springframework.boot.web.reactive.server.ConfigurableReactiveWebServerFactory
import org.springframework.context.support.GenericApplicationContext
import org.springframework.context.support.registerBean
import org.springframework.core.io.ClassPathResource
import org.springframework.fu.kofu.AbstractDsl
import org.springframework.fu.kofu.ConfigurationDsl
import org.springframework.fu.kofu.web.JacksonDsl
import org.springframework.web.reactive.function.server.CoRouterFunctionDsl
import org.springframework.web.reactive.function.server.RouterFunctionDsl
import org.springframework.web.server.WebFilter

/**
 * Kofu DSL for WebFlux server.
 *
 * This DSL to be used in [org.springframework.fu.kofu.application] and a
 * [org.springframework.boot.WebApplicationType.REACTIVE] parameter configures a
 * [WebFlux server](https://docs.spring.io/spring/docs/current/spring-framework-reference/webflux-reactive.html#spring-webflux).
 *
 * When no codec is configured, `String` and `Resource` ones are configured by default.
 * When a `codecs { }` block is declared, no one is configured by default.
 *
 * You can chose the underlying engine via the [WebFluxServerDsl.engine] parameter.
 *
 * Required dependencies can be retrieve using `org.springframework.boot:spring-boot-starter-webflux`.
 *
 * @sample org.springframework.fu.kofu.samples.webFluxCoRouter
 * @see org.springframework.fu.kofu.application
 * @see WebFluxServerDsl.codecs
 * @see WebFluxServerDsl.cors
 * @see WebFluxServerDsl.mustache
 * @author Sebastien Deleuze
 */
open class WebFluxServerDsl(private val init: WebFluxServerDsl.() -> Unit): AbstractDsl() {

	private val serverProperties = ServerProperties()

	private val resourceProperties = ResourceProperties()

	private val webFluxProperties = WebFluxProperties()

	private var codecsConfigured: Boolean = false

	/**
	 * Define the listening port of the webFlux.
	 */
	var port: Int = 8080

	/**
	 * Define the underlying engine used.
	 *
	 * @see netty
	 * @see tomcat
	 * @see jetty
	 * @see undertow
	 */
	var engine: ConfigurableReactiveWebServerFactory? = null

	override fun initialize(context: GenericApplicationContext) {
		super.initialize(context)
		init()
		context.registerBean(uniqueBeanName(RouterFunctionDsl::class.java.name, context)) {
			org.springframework.web.reactive.function.server.router {
				resources("/**", ClassPathResource("static/"))
			}
		}
		if (engine == null) {
			engine = netty()
		}
		engine!!.setPort(port)
		if (!codecsConfigured) {
			StringCodecInitializer(false, false).initialize(context)
			ResourceCodecInitializer(false).initialize(context)
		}
		if (context.containsBeanDefinition("webHandler")) {
			throw IllegalStateException("Only one webFlux per application is supported")
		}
		ReactiveWebServerInitializer(serverProperties, resourceProperties, webFluxProperties, engine).initialize(context)
	}

	/**
	 * Configure codecs via a [dedicated DSL][WebFluxServerCodecDsl].
	 */
	fun codecs(init: WebFluxServerCodecDsl.() -> Unit =  {}) {
		WebFluxServerCodecDsl(init).initialize(context)
		codecsConfigured = true
	}

	/**
	 * Define a request filter for this webFlux
	 */
	inline fun <reified T: WebFilter> filter() {
		context.registerBean<T>(uniqueBeanName(T::class.java.name, context))
	}

	/**
	 * Configure routes via a [dedicated DSL][RouterFunctionDsl].
	 * @sample org.springframework.fu.kofu.samples.webFluxRouter
	 */
	fun router(routes: (RouterFunctionDsl.() -> Unit)) {
		context.registerBean(uniqueBeanName(RouterFunctionDsl::class.java.name, context)) { org.springframework.web.reactive.function.server.router(routes) }
	}

	/**
	 * Get a reference to the bean by type or type + name with the syntax
	 * `ref<Foo>()` or `ref<Foo>("foo")`. When leveraging Kotlin type inference
	 * it could be as short as `ref()` or `ref("foo")`.
	 * TODO Update when SPR-17648 will be fixed
	 * @param name the name of the bean to retrieve
	 * @param T type the bean must match, can be an interface or superclass
	 */
	inline fun <reified T : Any> RouterFunctionDsl.ref(name: String? = null): T = when (name) {
		null -> context.getBean(T::class.java)
		else -> context.getBean(name, T::class.java)
	}

	/**
	 * Return an provider for the specified bean, allowing for lazy on-demand retrieval
	 * of instances, including availability and uniqueness options.
	 * TODO Update when SPR-17648 will be fixed
	 * @see org.springframework.beans.factory.BeanFactory.getBeanProvider
	 */
	inline fun <reified T : Any> RouterFunctionDsl.provider() : ObjectProvider<T> = context.getBeanProvider()

	/**
	 * Configure Coroutines routes via a [dedicated DSL][CoRouterFunctionDsl].
	 * @sample org.springframework.fu.kofu.samples.webFluxCoRouter
	 */
	fun coRouter(routes: (CoRouterFunctionDsl.() -> Unit)) {
		context.registerBean(uniqueBeanName(CoRouterFunctionDsl::class.java.name, context)) { org.springframework.web.reactive.function.server.coRouter(routes) }
	}

	/**
	 * Get a reference to the bean by type or type + name with the syntax
	 * `ref<Foo>()` or `ref<Foo>("foo")`. When leveraging Kotlin type inference
	 * it could be as short as `ref()` or `ref("foo")`.
	 * TODO Update when SPR-17648 will be fixed
	 * @param name the name of the bean to retrieve
	 * @param T type the bean must match, can be an interface or superclass
	 */
	inline fun <reified T : Any> CoRouterFunctionDsl.ref(name: String? = null): T = when (name) {
		null -> context.getBean(T::class.java)
		else -> context.getBean(name, T::class.java)
	}

	/**
	 * Return an provider for the specified bean, allowing for lazy on-demand retrieval
	 * of instances, including availability and uniqueness options.
	 * TODO Update when SPR-17648 will be fixed
	 * @see org.springframework.beans.factory.BeanFactory.getBeanProvider
	 */
	inline fun <reified T : Any> CoRouterFunctionDsl.provider() : ObjectProvider<T> = context.getBeanProvider()

	/**
	 * Netty engine.
	 * TODO Use lazy val when supported by GraalVM
	 * @see engine
	 */
	fun netty() = NettyDelegate().invoke()

	/**
	 * Tomcat engine.
	 * TODO Use lazy val when supported by GraalVM
	 * @see engine
	 */
	fun tomcat() =  TomcatDelegate().invoke()

	/**
	 * Jetty engine.
	 * TODO Use lazy val when supported by GraalVM
	 * @see engine
	 */
	fun jetty() = JettyDelegate().invoke()

	/**
	 * Undertow engine.
	 * TODO Use lazy val when supported by GraalVM
	 * @see engine
	 */
	fun undertow() = UndertowDelegate().invoke()


	private class NettyDelegate: () -> ConfigurableReactiveWebServerFactory {
		override fun invoke(): ConfigurableReactiveWebServerFactory {
			return NettyReactiveWebServerFactory()
		}
	}

	private class TomcatDelegate: () -> ConfigurableReactiveWebServerFactory {
		override fun invoke(): ConfigurableReactiveWebServerFactory {
			return TomcatReactiveWebServerFactory()
		}
	}

	private class JettyDelegate: () -> ConfigurableReactiveWebServerFactory {
		override fun invoke(): ConfigurableReactiveWebServerFactory {
			return JettyReactiveWebServerFactory()
		}
	}

	private class UndertowDelegate: () -> ConfigurableReactiveWebServerFactory {
		override fun invoke(): ConfigurableReactiveWebServerFactory {
			return UndertowReactiveWebServerFactory()
		}
	}

	class WebFluxServerCodecDsl(private val init: WebFluxServerCodecDsl.() -> Unit) : AbstractDsl() {

		override fun initialize(context: GenericApplicationContext) {
			super.initialize(context)
			init()
		}

		/**
		 * Enable [org.springframework.core.codec.CharSequenceEncoder] and [org.springframework.core.codec.StringDecoder]
		 */
		fun string(textPlainOnly: Boolean = false) {
			StringCodecInitializer(false, textPlainOnly).initialize(context)
		}

		/**
		 * Enable [org.springframework.http.codec.ResourceHttpMessageWriter] and [org.springframework.core.codec.ResourceDecoder]
		 */
		fun resource() {
			ResourceCodecInitializer(false).initialize(context)
		}

		/**
		 * Register an `ObjectMapper` bean and configure a [Jackson](https://github.com/FasterXML/jackson)
		 * JSON codec on WebFlux server via a [dedicated DSL][JacksonDsl].
		 *
		 * Required dependencies can be retrieve using `org.springframework.boot:spring-boot-starter-json`
		 * (included by default in `spring-boot-starter-webflux`).
		 *
		 * @sample org.springframework.fu.kofu.samples.jacksonDsl
		 */
		fun jackson(dsl: JacksonDsl.() -> Unit = {}) {
			JacksonDsl(dsl).initialize(context)
			JacksonJsonCodecInitializer(false).initialize(context)
		}

		/**
		 * Enable [org.springframework.http.codec.protobuf.ProtobufEncoder] and [org.springframework.http.codec.protobuf.ProtobufDecoder]
		 *
		 * This codec requires Protobuf 3 or higher with the official `com.google.protobuf:protobuf-java` dependency, and
		 * supports `application/x-protobuf` and `application/octet-stream`.
		 */
		fun protobuf() {
			ProtobufCodecInitializer(false).initialize(context)
		}

		/**
		 * Enable [org.springframework.http.codec.FormHttpMessageWriter] and [org.springframework.http.codec.FormHttpMessageReader]
		 */
		fun form() {
			FormCodecInitializer(false).initialize(context)
		}

		/**
		 * Enable [org.springframework.http.codec.multipart.MultipartHttpMessageWriter] and
		 * [org.springframework.http.codec.multipart.MultipartHttpMessageReader]
		 *
		 * This codec requires Synchronoss NIO Multipart library via  the `org.synchronoss.cloud:nio-multipart-parser`
		 * dependency.
		 */
		fun multipart() {
			MultipartCodecInitializer(false).initialize(context)
		}
	}

}

/**
 * Declare a WebFlux server.
 * @see WebFluxServerDsl
 */
fun ConfigurationDsl.webFlux(dsl: WebFluxServerDsl.() -> Unit =  {}) {
	WebFluxServerDsl(dsl).initialize(context)
}
